Adjacency analysis fails on fuel cell model with the NB

Please check [SOFCPoliMi.Tests.BenchmarkCammarata](https://libraries.openmodelica.org/branches/newInst-newBackend/SOFCPoliMi/files/SOFCPoliMi_SOFCPoliMi.Tests.BenchmarkCammarata.err). The NB fails with
```
Error: Internal error NBSlice.resolveDependency (single dependency) failed because list of scalar variables(10) {stack.module[$i1].pen.D.pCathode}, does not fit the equation size 100.
Error: Internal error NBSlice.resolveDependency failed for: stack.module[$i1].pen.D.pCathode.
Error: Internal error NBAdjacency.Matrix.upgradeRow failed for:
[FOR-] (100) ($RES_SIM_651)
[----] for $i1 in 1:10 loop
[----]   [ARRY] (10) stack.module[$i1].pen.D.D_O2_i = {(((5.959770780848035e-24 * sqrt(31.25117191894696 + 1/({0.01801528, 0.00201588, 0.0440095, 0.0280101, 0.01604246, 0.03006904, 0.04409562, 0.0319988, 0.0280134, 0.039948})[i]) * stack.module[$i1].pen.D.Temp ^ 1.5 * stack.module[$i1].pen.D.ratioPorTor) / stack.module[$i1].pen.D.pCathode) / ({3.054e-10, 3.147e-10, 3.7040000000000003e-10, 3.5785e-10, 3.6125e-10, 3.9550000000000003e-10, 4.2925000000000004e-10, 3.467e-10, 3.6325e-10, 3.5045e-10})[i] ^ 2.0) / stack.module[$i1].pen.D.collIntegral_O2_i[i] for i in 1:10} ($RES_SIM_652)
[----] end for;
```
Can you please check? Thanks!

Since in both cases we have some evidence of issue with the concLosses component, we tried to test that component standalone:
```modelica
model Test
  SOFCPoliMi.Components.FuelCell.DiffusionLosses.DiffusivityCoefficients diff(
    pAnode = 30e5+time,
    pCathode = 30e5-time,
    T = 1000+100*time,
    pXa = diff.pAnode*{0.4-0.1*time, 0.1, 0.2, 0.1, 0.15, 0.0, 0.0, 0.05 + 0.1*time, 0.0, 0.0},
    pXc = diff.pCathode*{0.0, 0.0, 0.7-0.1*time, 0.0, 0.0, 0.0, 0.0, 0.3+0.1*time, 0.0, 0.0}
    );
  annotation(
    uses(SOFCPolimi),
    __OpenModelica_commandLineOptions="--newBackend");
end Test;
```
The test works fine with the old backend. With the new one, it goes through the backend but then invalid C code is generated.

```C
Test_functions.c:31:24: error: passing 'void *' to parameter of incompatible type 'base_array_t' (aka 'struct base_array_s')
   31 |   real_array_copy_data(_OMC_LIT0, _table);
      |   ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
./Test_literals.h:7:19: note: expanded from macro '_OMC_LIT0'
    7 | #define _OMC_LIT0 MMC_REFSTRINGLIT(_OMC_LIT_STRUCT0)
      |                   ^
C:/Program Files/OpenModelica1.25.0-dev-64bit/include/omc/c/meta/meta_modelica_data.h:188:32: note: expanded from macro 'MMC_REFSTRINGLIT'
  188 | #define MMC_REFSTRINGLIT(NAME) MMC_TAGPTR(&(NAME).header)
      |                                ^
C:/Program Files/OpenModelica1.25.0-dev-64bit/include/omc/c/meta/meta_modelica_data.h:106:35: note: expanded from macro 'MMC_TAGPTR'
  106 | #define MMC_TAGPTR(p)             ((void*)((char*)(p) + 3))
      |                                   ^
C:/Program Files/OpenModelica1.25.0-dev-64bit/include/omc/c/util/generic_array.h:48:76: note: expanded from macro 'real_array_copy_data'
   48 | #define real_array_copy_data(src,dst)               simple_array_copy_data(src, &dst, sizeof(modelica_real));
      |                                                                            ^~~
C:/Program Files/OpenModelica1.25.0-dev-64bit/include/omc/c/util/generic_array.h:46:48: note: passing argument to parameter 'src' here
   46 | void simple_array_copy_data(const base_array_t src, base_array_t* dst, size_t sze);
      |                                                ^
```
By looking at the equation structure with the debugger, there are several expressions with indexing of arrays, such as
for i in 1:10 loop
```modelica
for i in 1:10 loop
    $FUN_3[i] := SOFCPoliMi.Components.FuelCell.DiffusionLosses.Utilities.CollisionIntegral(8, i, diff.Temp / {293.82132, 79.81222, 144.31854, 98.91608, 125.9191, 151.70758, 159.05524, 106.7, 87.28333, 99.7753}[i]);
  end for;
```
or
```modelica
for i in 1:10 loop
    diff.D_H2O_i[i] := 5.959770780848035e-24 * sqrt(55.508435061791985 + 1.0 / {0.01801528, 0.00201588, 0.0440095, 0.0280101, 0.01604246, 0.03006904, 0.04409562, 0.0319988, 0.0280134, 0.039948}[i]) * diff.Temp ^ 1.5 * diff.ratioPorTor / (3e6 + time) / {2.641e-10, 2.734e-10, 3.291e-10, 3.1655e-10, 3.1995e-10, 3.542e-10, 3.6595e-10, 3.054e-10, 3.2195e-10, 3.0915e-10}[i] ^ 2.0 / diff.collIntegral_H2O_i[i];
  end for;
```
not sure if they could be problematic for adjacency analysis in more complicated cases than this one, where we prescribe the input variables to be known time-varying expressions. Note that the latter solved equation comes from function inlining, when doing it with the OB, inlining is not applied
```
(assign) diff.D_H2O_i[1] := SOFCPoliMi.Components.FuelCell.DiffusionLosses.Utilities.DiffusivityEffCoeff(diff.Temp, diff.pAnode, 1, 1, 2.641e-10, diff.collIntegral_H2O_i[1], diff.ratioPorTor)
```
I'm not sure if this is the root cause of the failure of the adjacency analysis of the full model, but I guess it would be worth investigating why the generated code of Test is wrong, fix it, and make sure that the obtained results are the same with both backends.

--- a/OMCompiler/Compiler/NBackEnd/Classes/NBEquation.mo
+++ b/OMCompiler/Compiler/NBackEnd/Classes/NBEquation.mo
@@ -91,7 +91,7 @@ public
 
   // used to process different outcomes of slicing from Util/Slice.mo
   // have to be defined here and not in Util/Slice.mo because it is a uniontype and not a package
-  type Frame                = tuple<ComponentRef, Expression>                       "iterator-like tuple for array handling";
+  type Frame                = tuple<ComponentRef, Expression, Option<Iterator>>     "iterator-like tuple for array handling";
   type FrameLocation        = tuple<array<Integer>, Frame>                          "sliced frame at specific sub locations";
   type SlicingStatus        = enumeration(UNCHANGED, TRIVIAL, NONTRIVIAL, FAILURE)  "final result of slicing";
   type RecollectStatus      = enumeration(SUCCESS, FAILURE)                         "result of sub-routine recollect";
@@ -127,11 +127,13 @@ public
     record SINGLE
       ComponentRef name           "the name of the iterator";
       Expression range            "range as <start, step, stop>";
+      Option<Iterator> map        "maps to a second iterator if derived from a for-expression";
     end SINGLE;
 
     record NESTED
       array<ComponentRef> names   "sorted iterator names";
       array<Expression> ranges    "sorted ranges as <start, step, stop>";
+      array<Option<Iterator>> maps"maps to a second iterator if derived from a for-expression";
     end NESTED;
 
     record EMPTY
@@ -143,16 +145,18 @@ public
     protected
       list<ComponentRef> names;
       list<Expression> ranges;
+      list<Option<Iterator>> maps;
       ComponentRef name;
       Expression range;
+      Option<Iterator> map;
     algorithm
       if listEmpty(frames) then
         iter := EMPTY();
       else
-        (names, ranges) := List.unzip(frames);
-        iter := match (names, ranges)
-          case ({name}, {range})  then SINGLE(name, range);
-                                  else NESTED(listArray(names), listArray(ranges));
+        (names, ranges, maps) := List.unzip3(frames);
+        iter := match (names, ranges, maps)
+          case ({name}, {range}, {map})  then SINGLE(name, range, map);
+                                  else NESTED(listArray(names), listArray(ranges), listArray(maps));
         end match;
       end if;
     end fromFrames;
@@ -163,11 +167,12 @@ public
     protected
       list<ComponentRef> names1, names2;
       list<Expression> ranges1, ranges2;
+      list<Option<Iterator>> maps1, maps2;
     algorithm
       if not listEmpty(frames) then
-        (names1, ranges1) := getFrames(iter);
-        (names2, ranges2) := List.unzip(frames);
-        iter := fromFrames(List.zip(listAppend(names1, names2), listAppend(ranges1, ranges2)));
+        (names1, ranges1, maps1) := getFrames(iter);
+        (names2, ranges2, maps2) := List.unzip3(frames);
+        iter := fromFrames(List.zip3(listAppend(names1, names2), listAppend(ranges1, ranges2), listAppend(maps1, maps2)));
       end if;
     end addFrames;
 
@@ -175,11 +180,12 @@ public
       input Iterator iter;
       output list<ComponentRef> names;
       output list<Expression> ranges;
+      output list<Option<Iterator>> maps;
     algorithm
-      (names, ranges) := match iter
-        case SINGLE() then ({iter.name}, {iter.range});
-        case NESTED() then (arrayList(iter.names), arrayList(iter.ranges));
-        case EMPTY()  then ({}, {});
+      (names, ranges, maps) := match iter
+        case SINGLE() then ({iter.name}, {iter.range}, {iter.map});
+        case NESTED() then (arrayList(iter.names), arrayList(iter.ranges), arrayList(iter.maps));
+        case EMPTY()  then ({}, {}, {});
       end match;
     end getFrames;
 
@@ -190,16 +196,18 @@ public
     protected
       list<ComponentRef> tmp_names, names = {};
       list<Expression> tmp_ranges, ranges = {};
+      list<Option<Iterator>> tmp_maps, maps = {};
     algorithm
       if List.hasOneElement(iterators) then
         result := List.first(iterators);
       else
         for iter in listReverse(iterators) loop
-          (tmp_names, tmp_ranges) := getFrames(iter);
-          names := listAppend(tmp_names, names);
-          ranges := listAppend(tmp_ranges, ranges);
+          (tmp_names, tmp_ranges, tmp_maps) := getFrames(iter);
+          names   := listAppend(tmp_names, names);
+          ranges  := listAppend(tmp_ranges, ranges);
+          maps    := listAppend(tmp_maps, maps);
         end for;
-        result := NESTED(listArray(names), listArray(ranges));
+        result := NESTED(listArray(names), listArray(ranges), listArray(maps));
       end if;
     end merge;
 
@@ -211,9 +219,10 @@ public
     protected
       list<ComponentRef> names;
       list<Expression> ranges;
+      list<Option<Iterator>> maps;
     algorithm
-      (names, ranges) := getFrames(iterator);
-      for tpl in List.zip(names, ranges) loop
+      (names, ranges, maps) := getFrames(iterator);
+      for tpl in List.zip3(names, ranges, maps) loop
         result := Iterator.fromFrames({tpl}) :: result;
       end for;
     end split;
@@ -251,11 +260,15 @@ public
     algorithm
       b := match (iter1, iter2)
         case (EMPTY(), EMPTY()) then true;
-        case (SINGLE(), SINGLE()) then Expression.isEqual(iter1.range, iter2.range);
+        case (SINGLE(), SINGLE()) then Expression.isEqual(iter1.range, iter2.range) and Util.optionEqual(iter1.map, iter2.map, isEqual);
         case (NESTED(), NESTED()) algorithm
-          if arrayLength(iter1.ranges) == arrayLength(iter2.ranges) then
+          if arrayLength(iter1.ranges) == arrayLength(iter2.ranges) and arrayLength(iter1.maps) == arrayLength(iter2.maps) then
             for i in 1:arrayLength(iter1.ranges) loop
               b := Expression.isEqual(iter1.ranges[i], iter2.ranges[i]);
+              if not b then break; end if;
+            end for;
+            for i in 1:arrayLength(iter1.maps) loop
+              b := Util.optionEqual(iter1.maps[i], iter2.maps[i], isEqual);
               if not b then break; end if;
             end for;
           else
@@ -305,14 +318,13 @@ public
                   ty    = Expression.typeOf(iter1.range),
                   start = Expression.INTEGER(start_max),
                   step  = SOME(Expression.INTEGER(step1)),
-                  stop  = Expression.INTEGER(stop_min)
-                )
-              );
+                  stop  = Expression.INTEGER(stop_min)),
+                map  = iter1.map);
             end if;
 
             // create rest
-            rest1 := intersectRest(iter1.name, start1, step1, stop1, start_max-step1, stop_min+step1);
-            rest2 := intersectRest(iter2.name, start2, step2, stop2, start_max-step2, stop_min+step2);
+            rest1 := intersectRest(iter1.name, start1, step1, stop1, start_max-step1, stop_min+step1, iter1.map);
+            rest2 := intersectRest(iter2.name, start2, step2, stop2, start_max-step2, stop_min+step2, iter2.map);
         then (intersection, rest1, rest2);
 
         // cannot intersect
@@ -327,6 +339,7 @@ public
       input Integer stop;
       input Integer start_max;
       input Integer stop_min;
+      input Option<Iterator> map;
       output tuple<Iterator, Iterator> rest;
     protected
       Iterator rest_left, rest_right;
@@ -339,9 +352,8 @@ public
           range = Expression.makeRange(
             start = Expression.INTEGER(start),
             step  = SOME(Expression.INTEGER(step)),
-            stop  = Expression.INTEGER(start_max)
-          )
-        );
+            stop  = Expression.INTEGER(start_max)),
+          map   = map);
       end if;
 
       if stop_min > stop  then
@@ -352,9 +364,8 @@ public
           range = Expression.makeRange(
             start = Expression.INTEGER(stop_min),
             step  = SOME(Expression.INTEGER(step)),
-            stop  = Expression.INTEGER(stop)
-          )
-        );
+            stop  = Expression.INTEGER(stop)),
+          map   = map);
       end if;
       rest := (rest_left, rest_right);
     end intersectRest;
@@ -413,12 +424,14 @@ public
         case SINGLE() guard(arrayLength(location) == 1) algorithm
           (start, step, _) := Expression.getIntegerRange(iter.range);
           UnorderedMap.add(iter.name, Expression.INTEGER(start + location[1]*step), replacements);
+          createMappedLocationReplacement(iter.map, location[1], replacements);
         then ();
 
         case NESTED() guard(arrayLength(location) == arrayLength(iter.ranges)) algorithm
           for i in 1:arrayLength(location) loop
             (start, step, _) := Expression.getIntegerRange(iter.ranges[i]);
             UnorderedMap.add(iter.names[i], Expression.INTEGER(start + location[i]*step), replacements);
+            createMappedLocationReplacement(iter.maps[i], location[i], replacements);
           end for;
         then ();
 
@@ -428,6 +441,27 @@ public
         then fail();
       end match;
     end createLocationReplacements;
+
+    function createMappedLocationReplacement
+      input Option<Iterator> map;
+      input Integer location;
+      input UnorderedMap<ComponentRef, Expression> replacements   "replacement rules";
+    algorithm
+      _ := match map
+        local
+          ComponentRef name;
+          Expression arr;
+          Integer index;
+
+        // only does something if the option is filled with an array
+        // fail if there is something else?
+        case SOME(SINGLE(name = name, range = arr as Expression.ARRAY())) algorithm
+          index := Expression.getInteger(arr.elements[location]);
+          UnorderedMap.add(name, Expression.INTEGER(index), replacements);
+        then ();
+        else ();
+      end match;
+    end createMappedLocationReplacement;
 
     function createReplacement
       "adds a replacement rule for one iterator to another.
@@ -529,7 +563,7 @@ public
         case Expression.CALL(call = call as Call.TYPED_ARRAY_CONSTRUCTOR()) algorithm
           for tpl in listReverse(call.iters) loop
             (node, range) := tpl;
-            frames := (ComponentRef.fromNode(node, Type.INTEGER(), {}, NFComponentRef.Origin.ITERATOR), range) :: frames;
+            frames := (ComponentRef.fromNode(node, Type.INTEGER(), {}, NFComponentRef.Origin.ITERATOR), range, NONE()) :: frames;
           end for;
           tmp := fromFrames(frames);
           if not isEmpty(iter) then
@@ -604,12 +638,20 @@ public
       function singleStr
         input ComponentRef name;
         input Expression range;
+        input Option<Iterator> map;
         output String str = ComponentRef.toString(name) + " in " + Expression.toString(range);
+      protected
+        list<ComponentRef> names;
+      algorithm
+        if Util.isSome(map) then
+          (names, _) := getFrames(Util.getOption(map));
+          str := str + " (" + ComponentRef.toString(List.first(names)) + ")";
+        end if;
       end singleStr;
     algorithm
       str := match iter
-        case SINGLE() then singleStr(iter.name, iter.range);
-        case NESTED() then "{" + stringDelimitList(list(singleStr(iter.names[i], iter.ranges[i]) for i in 1:arrayLength(iter.names)), ", ") + "}";
+        case SINGLE() then singleStr(iter.name, iter.range, iter.map);
+        case NESTED() then "{" + stringDelimitList(list(singleStr(iter.names[i], iter.ranges[i], iter.maps[i]) for i in 1:arrayLength(iter.names)), ", ") + "}";
         case EMPTY()  then "<EMPTY ITERATOR>";
         else algorithm
           Error.addMessage(Error.INTERNAL_ERROR,{getInstanceName() + " failed for an unknown reason."});
@@ -1715,10 +1757,11 @@ public
         local
           list<ComponentRef> names;
           list<Expression> ranges;
+          list<Option<Iterator>> maps;
 
         case FOR_EQUATION() algorithm
-          (names, ranges) := Iterator.getFrames(eqn.iter);
-        then List.zip(names, ranges);
+          (names, ranges, maps) := Iterator.getFrames(eqn.iter);
+        then List.zip3(names, ranges, maps);
 
         else {};
       end match;

--- a/OMCompiler/Compiler/NBackEnd/Classes/NBackendDAE.mo
+++ b/OMCompiler/Compiler/NBackEnd/Classes/NBackendDAE.mo
@@ -895,7 +895,7 @@ protected
             body_elem := Pointer.access(body_elem_ptr);
             body_elem := BEquation.FOR_EQUATION(
               size    = Expression.rangeSize(range) * Equation.size(body_elem_ptr),
-              iter    = Iterator.SINGLE(iterator, range),
+              iter    = Iterator.SINGLE(iterator, range, NONE()),
               body    = {body_elem},
               source  = frontend_equation.source,
               attr    = Equation.getAttributes(body_elem)

--- a/OMCompiler/Compiler/NBackEnd/Modules/1_Main/NBInitialization.mo
+++ b/OMCompiler/Compiler/NBackEnd/Modules/1_Main/NBInitialization.mo
@@ -405,7 +405,7 @@ public
         Expression e;
         Call array_constructor;
         list<Subscript> subscripts;
-        list<tuple<ComponentRef, Expression>> frames;
+        list<tuple<ComponentRef, Expression, Option<Iterator>>> frames;
         UnorderedMap<ComponentRef, Expression> replacements;
         InstNode old_iter;
         ComponentRef new_iter;
@@ -415,7 +415,7 @@ public
         (var_ptr, name, _, _, _, frames, iterator) := createIteratedStartCref(var_ptr, name);
         replacements := UnorderedMap.new<Expression>(ComponentRef.hash, ComponentRef.isEqual);
         for tpl in List.zip(array_constructor.iters, frames) loop
-          ((old_iter, _), (new_iter, _)) := tpl;
+          ((old_iter, _), (new_iter, _, _)) := tpl;
           UnorderedMap.add(ComponentRef.fromNode(old_iter, InstNode.getType(old_iter)), Expression.fromCref(new_iter), replacements);
         end for;
       then Expression.map(array_constructor.exp, function Replacements.applySimpleExp(replacements = replacements));
@@ -460,7 +460,7 @@ public
     output Pointer<Variable> start_var;
     output ComponentRef start_cref;
     output list<Subscript> subscripts;
-    output list<tuple<ComponentRef, Expression>> frames;
+    output list<tuple<ComponentRef, Expression, Option<Iterator>>> frames;
     output Iterator iterator;
   protected
     list<Dimension> dims;
@@ -474,7 +474,7 @@ public
     iter_crefs  := list(ComponentRef.makeIterator(iter, Type.INTEGER()) for iter in iterators);
     iter_crefs  := list(BackendDAE.lowerIteratorCref(iter) for iter in iter_crefs);
     subscripts  := list(Subscript.mapExp(sub, BackendDAE.lowerIteratorExp) for sub in subscripts);
-    frames      := List.zip(iter_crefs, ranges);
+    frames      := List.zip3(iter_crefs, ranges, List.fill(NONE(), listLength(iter_crefs)));
     iterator    := Iterator.fromFrames(frames);
 
     // create start variable name with subscripts and create start expression
@@ -515,7 +515,7 @@ public
     list<InstNode> iterators;
     list<Expression> ranges;
     list<Subscript> subscripts;
-    list<tuple<ComponentRef, Expression>> frames;
+    list<tuple<ComponentRef, Expression, Option<Iterator>>> frames;
     Pointer<Equation> pre_eq;
     EquationKind kind;
   algorithm
@@ -526,7 +526,7 @@ public
         name    := BVariable.getVarName(var_ptr);
         dims    := Type.arrayDims(ComponentRef.getSubscriptedType(name));
         (iterators, ranges, subscripts) := Flatten.makeIterators(name, dims);
-        frames  := List.zip(list(ComponentRef.makeIterator(iter, Type.INTEGER()) for iter in iterators), ranges);
+        frames  := List.zip3(list(ComponentRef.makeIterator(iter, Type.INTEGER()) for iter in iterators), ranges, List.fill(NONE(), listLength(ranges)));
 
         pre_name := BVariable.getVarName(Util.getOption(pre));
         pre_name := ComponentRef.mergeSubscripts(subscripts, pre_name, true, true);

--- a/OMCompiler/Compiler/NBackEnd/Modules/2_Pre/NBEvents.mo
+++ b/OMCompiler/Compiler/NBackEnd/Modules/2_Pre/NBEvents.mo
@@ -221,15 +221,16 @@ public
     protected
       list<ComponentRef> iter;
       list<Expression> range;
+      list<Option<Iterator>> maps;
       ComponentRef lhs_cref;
       Pointer<Equation> aux_eqn;
     algorithm
       // if it has a statement index, it already has been created as a statement inside an algorithm (0 implies no index)
       if cond.stmt_index == 0 then
-        (iter, range) := Equation.Iterator.getFrames(cond.iter);
+        (iter, range, maps) := Equation.Iterator.getFrames(cond.iter);
         // lower the subscripts (containing iterators)
         lhs_cref := ComponentRef.mapSubscripts(BVariable.getVarName(aux_var), function Subscript.mapExp(func = function BackendDAE.lowerComponentReferenceExp(variables = variables)));
-        aux_eqn := Equation.makeAssignment(Expression.fromCref(lhs_cref), cond.exp, idx, "EVT", Iterator.fromFrames(List.zip(iter, range)), EquationAttributes.default(EquationKind.DISCRETE, false));
+        aux_eqn := Equation.makeAssignment(Expression.fromCref(lhs_cref), cond.exp, idx, "EVT", Iterator.fromFrames(List.zip3(iter, range, maps)), EquationAttributes.default(EquationKind.DISCRETE, false));
         auxiliary_eqns := aux_eqn :: auxiliary_eqns;
       end if;
       // remove all subscripts from the variable name
@@ -636,7 +637,7 @@ public
         case Statement.FOR(range = SOME(range)) algorithm
           new_stmts := {};
           name := ComponentRef.fromNode(stmt.iterator, Type.INTEGER());
-          new_frames := (name, range) :: frames;
+          new_frames := (name, range, NONE()) :: frames;
           for elem in stmt.body loop
             new_stmt := fromStatement(elem, bucket_ptr, eqn, variables, funcTree, new_frames);
             new_stmts := new_stmt :: new_stmts;

--- a/OMCompiler/Compiler/NBackEnd/Modules/2_Pre/NBFunctionAlias.mo
+++ b/OMCompiler/Compiler/NBackEnd/Modules/2_Pre/NBFunctionAlias.mo
@@ -59,7 +59,7 @@ protected
   import Partitioning = NBPartitioning;
   import NBPartitioning.BClock;
   import BVariable = NBVariable;
-  import NBVariable.{VariablePointers, VarData};
+  import NBVariable.{VariablePointer, VariablePointers, VarData};
 
   // Util imports
   import StringUtil;
@@ -221,6 +221,7 @@ protected
   protected
     UnorderedMap<Call_Id, Call_Aux> map = UnorderedMap.new<Call_Aux>(Call_Id.hash, Call_Id.isEqual);
     VariablePointers variables = VarData.getVariables(varData);
+    UnorderedSet<VariablePointer> set = UnorderedSet.new(BVariable.hash, BVariable.equalName) "new iterators";
     UnorderedMap<BClock, ComponentRef> clock_map;
     Pointer<Integer> aux_index = Pointer.create(1);
     list<Pointer<Variable>> new_vars_disc = {}, new_vars_cont = {}, new_vars_init = {}, new_vars_recd = {}, new_vars_clck;
@@ -240,7 +241,7 @@ protected
       case EqData.EQ_DATA_SIM() algorithm
         // first collect all new functions from simulation equations
         eqData.simulation := EquationPointers.map(eqData.simulation,
-          function introduceFunctionAliasEquation(map = map, variables = variables, aux_index = aux_index, eqn_index = eqData.uniqueIndex, init = false));
+          function introduceFunctionAliasEquation(map = map, variables = variables, set = set, aux_index = aux_index, eqn_index = eqData.uniqueIndex, init = false));
 
         // create new simulation variables and corresponding equations for the function alias
         for tpl in listReverse(UnorderedMap.toList(map)) loop
@@ -271,7 +272,7 @@ protected
 
         // afterwards collect all functions from initial equations
         eqData.initials := EquationPointers.map(eqData.initials,
-          function introduceFunctionAliasEquation(map = map, variables = variables, aux_index = aux_index, eqn_index = eqData.uniqueIndex, init = true));
+          function introduceFunctionAliasEquation(map = map, variables = variables, set = set, aux_index = aux_index, eqn_index = eqData.uniqueIndex, init = true));
 
         // create new initialization variables and corresponding equations for the function alias
         for tpl in listReverse(UnorderedMap.toList(map)) loop
@@ -302,6 +303,7 @@ protected
     varData := VarData.addTypedList(varData, new_vars_init, VarData.VarType.PARAMETER);
     varData := VarData.addTypedList(varData, new_vars_recd, VarData.VarType.RECORD);
     varData := VarData.addTypedList(varData, new_vars_clck, VarData.VarType.CLOCK);
+    varData := VarData.addTypedList(varData, UnorderedSet.toList(set), VarData.VarType.ITERATOR);
     eqData  := EqData.addTypedList(eqData, new_eqns_cont, EqData.EqType.CONTINUOUS, false);
     eqData  := EqData.addTypedList(eqData, new_eqns_disc, EqData.EqType.DISCRETE, false);
     eqData  := EqData.addTypedList(eqData, new_eqns_init, EqData.EqType.INITIAL, false);
@@ -334,6 +336,7 @@ protected
     input output Equation eqn;
     input UnorderedMap<Call_Id, Call_Aux> map;
     input VariablePointers variables;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Pointer<Integer> aux_index;
     input Pointer<Integer> eqn_index;
     input Boolean init;
@@ -342,7 +345,7 @@ protected
     Boolean stop;
   algorithm
     // inline trivial array constructors first
-    eqn := Inline.inlineArrayConstructorSingle(eqn, Iterator.EMPTY(), variables, eqn_index);
+    eqn := Inline.inlineArrayConstructorSingle(eqn, Iterator.EMPTY(), variables, set, eqn_index);
 
     // get iterator and determine if it needs to be checked further
     (iter, stop) := match eqn
@@ -376,6 +379,7 @@ protected
       local
         list<ComponentRef> names;
         list<Expression> ranges;
+        list<Option<Iterator>> maps;
         Iterator new_iter;
         Call_Id id;
         ComponentRef name;
@@ -383,14 +387,13 @@ protected
         Call_Aux aux;
         Option<Call_Aux> aux_opt;
         Expression new_exp, sub_exp;
-        list<ComponentRef> names;
-        list<Expression> tpl_lst;
+       list<Expression> tpl_lst;
 
       case Expression.CALL() guard(checkCallReplacement(exp.call)) algorithm
         // strip nested iterator for the iterators that actually occure in the function call
         if not Iterator.isEmpty(iter) then
-          (names, ranges) := Iterator.getFrames(iter);
-          new_iter := Iterator.fromFrames(filterFrames(exp, names, ranges));
+          (names, ranges, maps) := Iterator.getFrames(iter);
+          new_iter := Iterator.fromFrames(filterFrames(exp, names, ranges, maps));
         else
           new_iter := iter;
         end if;
@@ -486,7 +489,8 @@ protected
     input Expression exp;
     input list<ComponentRef> names;
     input list<Expression> ranges;
-    output list<tuple<ComponentRef, Expression>> frames;
+    input list<Option<Iterator>> maps;
+    output list<tuple<ComponentRef, Expression, Option<Iterator>>> frames;
   protected
     UnorderedMap<ComponentRef, Expression> frame_map = UnorderedMap.fromLists<Expression>(names, ranges, ComponentRef.hash, ComponentRef.isEqual);
     UnorderedMap<ComponentRef, Expression> new_map = UnorderedMap.new<Expression>(ComponentRef.hash, ComponentRef.isEqual);
@@ -510,9 +514,16 @@ protected
         else ();
       end match;
     end collectFrames;
+
+    list<ComponentRef> n;
+    list<Expression> r;
+    list<Option<Iterator>> m;
   algorithm
     _ := Expression.map(exp, function collectFrames(frame_map = frame_map, new_map = new_map));
-    frames := UnorderedMap.toList(new_map);
+    n := UnorderedMap.keyList(new_map);
+    r := UnorderedMap.valueList(new_map);
+    m := List.fill(NONE(), listLength(n));
+    frames := List.zip3(n, r, m);
   end filterFrames;
 
   function addAuxVar

--- a/OMCompiler/Compiler/NBackEnd/Modules/2_Pre/NBInline.mo
+++ b/OMCompiler/Compiler/NBackEnd/Modules/2_Pre/NBInline.mo
@@ -54,6 +54,7 @@ protected
   import Statement = NFStatement;
   import Subscript = NFSubscript;
   import Type = NFType;
+  import Variable = NFVariable;
 
   // NB imports
   import Module = NBModule;
@@ -159,13 +160,14 @@ public
   function inlineRecordSliceEquation
     input Slice<Pointer<Equation>> slice;
     input VariablePointers variables;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Pointer<Integer> index;
     input Boolean inlineSimple;
     output list<Slice<Pointer<Equation>>> slices;
   protected
     Pointer<list<Pointer<Equation>>> record_eqns = Pointer.create({});
   algorithm
-    inlineRecordTupleArrayEquation(Pointer.access(Slice.getT(slice)), Iterator.EMPTY(), variables, record_eqns, index, inlineSimple);
+    inlineRecordTupleArrayEquation(Pointer.access(Slice.getT(slice)), Iterator.EMPTY(), variables, record_eqns, set, index, inlineSimple);
     // somehow split slice.indices
     slices := list(Slice.SLICE(eqn, {}) for eqn in Pointer.access(record_eqns));
     if listEmpty(slices) then
@@ -177,6 +179,7 @@ public
     input output Equation eqn;
     input Iterator iter;
     input VariablePointers variables;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Pointer<Integer> index;
     input Pointer<list<Pointer<Equation>>> new_eqns = Pointer.create({});
     output Boolean changed;
@@ -190,15 +193,15 @@ public
 
         // CREF = {... for i in []} array constructor equation
         case Equation.ARRAY_EQUATION(lhs = lhs as Expression.CREF(), rhs = rhs as Expression.CALL(call = call as Call.TYPED_ARRAY_CONSTRUCTOR())) algorithm
-        then (inlineArrayConstructor(eqn, lhs.cref, call.exp, call.iters, eqn.attr, iter, variables, new_eqns, index), true);
+        then (inlineArrayConstructor(eqn, lhs.cref, call.exp, call.iters, eqn.attr, iter, variables, new_eqns, set, index), true);
 
         // {... for i in []} = CREF array constructor equation
         case Equation.ARRAY_EQUATION(lhs = lhs as Expression.CALL(call = call as Call.TYPED_ARRAY_CONSTRUCTOR()), rhs = rhs as Expression.CREF()) algorithm
-        then (inlineArrayConstructor(eqn, rhs.cref, call.exp, call.iters, eqn.attr, iter, variables, new_eqns, index), true);
+        then (inlineArrayConstructor(eqn, rhs.cref, call.exp, call.iters, eqn.attr, iter, variables, new_eqns, set, index), true);
 
         // apply on for-equation. assumed to be split up
         case Equation.FOR_EQUATION(body = {body}) algorithm
-          (new_eqn, changed) := inlineArrayConstructorSingle(body, eqn.iter, variables, index, new_eqns);
+          (new_eqn, changed) := inlineArrayConstructorSingle(body, eqn.iter, variables, set, index, new_eqns);
           new_eqn := if changed then new_eqn else eqn;
         then (new_eqn, changed);
 
@@ -231,14 +234,14 @@ protected
 
     // do not inline records tuples and arrays after causalization
     // ToDo: do it properly on strong components, cannot remove equations here
+    set     := UnorderedSet.new(BVariable.hash, BVariable.equalName);
     if not List.any(inline_types, function DAEUtil.inlineTypeEqual(it2 = DAE.AFTER_INDEX_RED_INLINE())) then
       // replace record constucters after functions because record operator
       // functions will produce record constructors once inlined
-      eqData  := inlineRecordsTuplesArrays(eqData, variables, init);
+      eqData  := inlineRecordsTuplesArrays(eqData, variables, set, init);
     end if;
 
     // collect new iterators from replaced function bodies
-    set     := UnorderedSet.new(BVariable.hash, BVariable.equalName);
     eqData  := EqData.map(eqData, function BackendDAE.lowerEquationIterators(variables = variables, set = set));
     varData := VarData.addTypedList(varData, UnorderedSet.toList(set), NBVariable.VarData.VarType.ITERATOR);
     eqData  := EqData.mapExp(eqData, function BackendDAE.lowerComponentReferenceExp(variables = variables));
@@ -262,6 +265,7 @@ protected
     "does not inline simple record equalities"
     input output EqData eqData;
     input VariablePointers variables;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Boolean init;
   protected
     Pointer<Integer> index = EqData.getUniqueIndex(eqData);
@@ -270,7 +274,7 @@ protected
     if init then
       eqData := match eqData
         case EqData.EQ_DATA_SIM() algorithm
-          eqData.initials := EquationPointers.map(eqData.initials, function inlineRecordTupleArrayEquation(iter = Iterator.EMPTY(), variables = variables, new_eqns = new_eqns, index = index, inlineSimple = false));
+          eqData.initials := EquationPointers.map(eqData.initials, function inlineRecordTupleArrayEquation(iter = Iterator.EMPTY(), variables = variables, new_eqns = new_eqns, set = set, index = index, inlineSimple = false));
           eqData.initials := EquationPointers.addList(Pointer.access(new_eqns), eqData.initials);
           eqData.initials := EquationPointers.compress(eqData.initials);
         then eqData;
@@ -278,7 +282,7 @@ protected
         else eqData;
       end match;
     else
-      eqData := EqData.map(eqData, function inlineRecordTupleArrayEquation(iter = Iterator.EMPTY(), variables = variables, new_eqns = new_eqns, index = index, inlineSimple = false));
+      eqData := EqData.map(eqData, function inlineRecordTupleArrayEquation(iter = Iterator.EMPTY(), variables = variables, new_eqns = new_eqns, set = set, index = index, inlineSimple = false));
       eqData := EqData.addUntypedList(eqData, Pointer.access(new_eqns), false);
       eqData := EqData.compress(eqData);
     end if;
@@ -289,6 +293,7 @@ protected
     input Iterator iter;
     input VariablePointers variables;
     input Pointer<list<Pointer<Equation>>> new_eqns;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Pointer<Integer> index;
     input Boolean inlineSimple;
   algorithm
@@ -306,45 +311,45 @@ protected
         case Equation.ARRAY_EQUATION(lhs = Expression.CREF(), rhs = Expression.CREF())  guard(not inlineSimple) then eqn;
 
         // try to inline other record equations. try catch to be sure to not discard
-        case Equation.RECORD_EQUATION(ty = Type.COMPLEX()) then inlineRecordEquation(eqn, eqn.lhs, eqn.rhs, iter, eqn.attr, eqn.recordSize, variables, new_eqns, index, inlineSimple);
+        case Equation.RECORD_EQUATION(ty = Type.COMPLEX()) then inlineRecordEquation(eqn, eqn.lhs, eqn.rhs, iter, eqn.attr, eqn.recordSize, variables, new_eqns, set, index, inlineSimple);
 
         // only if record size is not NONE()
-        case Equation.ARRAY_EQUATION(recordSize = SOME(size)) then inlineRecordEquation(eqn, eqn.lhs, eqn.rhs, iter, eqn.attr, size, variables, new_eqns, index, inlineSimple);
+        case Equation.ARRAY_EQUATION(recordSize = SOME(size)) then inlineRecordEquation(eqn, eqn.lhs, eqn.rhs, iter, eqn.attr, size, variables, new_eqns, set, index, inlineSimple);
 
         // inlining potential tuple equations
-        case Equation.RECORD_EQUATION() then inlineTupleEquation(eqn, eqn.lhs, eqn.rhs, eqn.attr, iter, variables, new_eqns, index);
+        case Equation.RECORD_EQUATION() then inlineTupleEquation(eqn, eqn.lhs, eqn.rhs, eqn.attr, iter, variables, new_eqns, set, index);
 
         // {...} = {...} array equation
         case Equation.ARRAY_EQUATION(lhs = lhs as Expression.ARRAY(), rhs = rhs as Expression.ARRAY())
-        then inlineArrayEquation(eqn, lhs.elements, rhs.elements, eqn.attr, iter, variables, new_eqns, index);
+        then inlineArrayEquation(eqn, lhs.elements, rhs.elements, eqn.attr, iter, variables, new_eqns, set, index);
 
         // CREF = {...} array equation
         case Equation.ARRAY_EQUATION(lhs = lhs as Expression.CREF(), rhs = rhs as Expression.ARRAY()) algorithm
           elements := list(NFExpression.applySubscripts({Subscript.INDEX(Expression.INTEGER(i))}, lhs) for i in 1:arrayLength(rhs.elements));
-        then inlineArrayEquation(eqn, listArray(elements), rhs.elements, eqn.attr, iter, variables, new_eqns, index);
+        then inlineArrayEquation(eqn, listArray(elements), rhs.elements, eqn.attr, iter, variables, new_eqns, set, index);
 
         // {...} = CREF array equation
         case Equation.ARRAY_EQUATION(lhs = lhs as Expression.ARRAY(), rhs = rhs as Expression.CREF()) algorithm
           elements := list(NFExpression.applySubscripts({Subscript.INDEX(Expression.INTEGER(i))}, rhs) for i in 1:arrayLength(lhs.elements));
-        then inlineArrayEquation(eqn, lhs.elements, listArray(elements), eqn.attr, iter, variables, new_eqns, index);
+        then inlineArrayEquation(eqn, lhs.elements, listArray(elements), eqn.attr, iter, variables, new_eqns, set, index);
 
         // CREF = {... for i in []} array constructor equation
         case Equation.ARRAY_EQUATION(lhs = lhs as Expression.CREF(), rhs = rhs as Expression.CALL(call = call as Call.TYPED_ARRAY_CONSTRUCTOR()))
-        then inlineArrayConstructor(eqn, lhs.cref, call.exp, call.iters, eqn.attr, iter, variables, new_eqns, index);
+        then inlineArrayConstructor(eqn, lhs.cref, call.exp, call.iters, eqn.attr, iter, variables, new_eqns, set, index);
 
         // {... for i in []} = CREF array constructor equation
         case Equation.ARRAY_EQUATION(lhs = lhs as Expression.CALL(call = call as Call.TYPED_ARRAY_CONSTRUCTOR()), rhs = rhs as Expression.CREF())
-        then inlineArrayConstructor(eqn, rhs.cref, call.exp, call.iters, eqn.attr, iter, variables, new_eqns, index);
+        then inlineArrayConstructor(eqn, rhs.cref, call.exp, call.iters, eqn.attr, iter, variables, new_eqns, set, index);
 
         // apply on for-equation. assumed to be split up
         case Equation.FOR_EQUATION(body = {body}) algorithm
-          new_eqn := inlineRecordTupleArrayEquation(body, eqn.iter, variables, new_eqns, index, true);
+          new_eqn := inlineRecordTupleArrayEquation(body, eqn.iter, variables, new_eqns, set, index, true);
           new_eqn := if Equation.isDummy(new_eqn) then new_eqn else eqn;
         then new_eqn;
 
         // apply on if-equation body equations
         case Equation.IF_EQUATION() guard IfEquationBody.isRecordOrTupleEquation(eqn.body) algorithm
-          new_eqn := inlineRecordTupleArrayIfEquation(eqn, eqn.body, iter, variables, new_eqns, index, inlineSimple);
+          new_eqn := inlineRecordTupleArrayIfEquation(eqn, eqn.body, iter, variables, new_eqns, set, index, inlineSimple);
           new_eqn := if Equation.isDummy(new_eqn) then new_eqn else eqn;
         then new_eqn;
 
@@ -365,6 +370,7 @@ protected
     input Iterator iter;
     input VariablePointers variables;
     input Pointer<list<Pointer<Equation>>> new_eqns;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Pointer<Integer> index;
     input Boolean inlineSimple;
   protected
@@ -373,7 +379,7 @@ protected
     Pointer<Equation> new_eqn;
   algorithm
     eqns := Pointer.access(new_eqns);
-    new_body := inlineRecordTupleArrayIfBody(body, iter, variables, index, inlineSimple);
+    new_body := inlineRecordTupleArrayIfBody(body, iter, variables, set, index, inlineSimple);
     for b in IfEquationBody.split(new_body) loop
       new_eqn := IfEquationBody.makeIfEquation(b, index, NBEquation.SIMULATION_STR, iter, Equation.getSource(eqn), Equation.getAttributes(eqn));
       eqns := new_eqn :: eqns;
@@ -387,17 +393,18 @@ protected
     input output IfEquationBody body;
     input Iterator iter;
     input VariablePointers variables;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Pointer<Integer> index;
     input Boolean inlineSimple;
   protected
     Pointer<list<Pointer<Equation>>> new_eqns = Pointer.create({});
   algorithm
     body.then_eqns := List.flatten(list(
-      match inlineRecordTupleArrayEquation(Pointer.access(e), iter, variables, new_eqns, index, inlineSimple)
+      match inlineRecordTupleArrayEquation(Pointer.access(e), iter, variables, new_eqns, set, index, inlineSimple)
         case Equation.DUMMY_EQUATION() then Pointer.access(new_eqns);
         else {e};
       end match for e in body.then_eqns));
-    body.else_if := Util.applyOption(body.else_if, function inlineRecordTupleArrayIfBody(iter = iter, variables = variables, index = index, inlineSimple = inlineSimple));
+    body.else_if := Util.applyOption(body.else_if, function inlineRecordTupleArrayIfBody(iter = iter, variables = variables, set = set, index = index, inlineSimple = inlineSimple));
   end inlineRecordTupleArrayIfBody;
 
   function inlineRecordEquation
@@ -412,6 +419,7 @@ protected
     input Integer recordSize;
     input VariablePointers variables;
     input Pointer<list<Pointer<Equation>>> new_eqns;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Pointer<Integer> index;
     input Boolean inlineSimple;
   protected
@@ -425,7 +433,7 @@ protected
     for i in 1:recordSize loop
       new_lhs := inlineRecordConstructorExp(lhs, i, variables);
       new_rhs := inlineRecordConstructorExp(rhs, i, variables);
-      eqns    := createInlinedEquation(eqns, new_lhs, new_rhs, attr, iter, variables, index);
+      eqns    := createInlinedEquation(eqns, new_lhs, new_rhs, attr, iter, variables, set, index);
     end for;
     Pointer.update(new_eqns, eqns);
     eqn := Equation.DUMMY_EQUATION();
@@ -440,6 +448,7 @@ protected
     input Iterator iter;
     input VariablePointers variables;
     input Pointer<list<Pointer<Equation>>> new_eqns;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Pointer<Integer> index;
   protected
     list<Pointer<Equation>> eqns;
@@ -458,7 +467,7 @@ protected
         (lhs, rhs) := tpl;
         // skip wild cref assignments
         if not (Expression.isWildCref(lhs) or Expression.isWildCref(rhs)) then
-          eqns := createInlinedEquation(eqns, lhs, rhs, attr, iter, variables, index);
+          eqns := createInlinedEquation(eqns, lhs, rhs, attr, iter, variables, set, index);
         end if;
       end for;
       Pointer.update(new_eqns, eqns);
@@ -477,6 +486,7 @@ protected
     input Iterator iter;
     input VariablePointers variables;
     input Pointer<list<Pointer<Equation>>> new_eqns;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Pointer<Integer> index;
   protected
     Pointer<list<Pointer<Equation>>> tmp_eqns;
@@ -489,7 +499,7 @@ protected
     end if;
     eqns := Pointer.access(new_eqns);
     for i in 1: arrayLength(lhs_elements) loop
-      eqns := createInlinedEquation(eqns, lhs_elements[i], rhs_elements[i], attr, iter, variables, index);
+      eqns := createInlinedEquation(eqns, lhs_elements[i], rhs_elements[i], attr, iter, variables, set, index);
     end for;
     Pointer.update(new_eqns, eqns);
     eqn := Equation.DUMMY_EQUATION();
@@ -504,10 +514,10 @@ protected
     input Iterator iter;
     input VariablePointers variables;
     input Pointer<list<Pointer<Equation>>> new_eqns;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Pointer<Integer> index;
   protected
-    list<tuple<ComponentRef, Expression>> frames
-      = list((ComponentRef.makeIterator(Util.tuple21(tpl), Type.INTEGER()), Util.tuple22(tpl)) for tpl in iters);
+    list<tuple<ComponentRef, Expression, Option<Iterator>>> frames;
     list<Subscript> subs;
     Expression cref_exp;
     list<Pointer<Equation>> eqns;
@@ -516,13 +526,54 @@ protected
       print("[" + getInstanceName() + "] Inlining: " + Equation.toString(eqn) + "\n");
     end if;
     eqns := Pointer.access(new_eqns);
+
+    // inline the iterators
+    frames := list(inlineArrayIterator(iter, set) for iter in iters);
+
     // add the iterators to the cref
-    subs      := list(Subscript.INDEX(Expression.CREF(Type.INTEGER(), Util.tuple21(tpl))) for tpl in frames);
+    subs      := Iterator.normalizedSubscripts(Iterator.fromFrames(frames));
+    //subs      := list(Subscript.INDEX(Expression.CREF(Type.INTEGER(), Util.tuple31(tpl))) for tpl in frames);
     cref_exp  := Expression.fromCref(ComponentRef.mergeSubscripts(subs, cref, true));
-    eqns      := createInlinedEquation(eqns, cref_exp, rhs, attr, Iterator.addFrames(iter, frames), variables, index);
+    eqns      := createInlinedEquation(eqns, cref_exp, rhs, attr, Iterator.addFrames(iter, frames), variables, set, index);
     Pointer.update(new_eqns, eqns);
     eqn := Equation.DUMMY_EQUATION();
   end inlineArrayConstructor;
+
+  function inlineArrayIterator
+    input tuple<InstNode, Expression> iter;
+    input UnorderedSet<VariablePointer> set "new iterators";
+    output tuple<ComponentRef, Expression, Option<Iterator>> frame;
+  algorithm
+    frame := match iter
+      local
+        InstNode node, node2;
+        Expression range, range2;
+        Iterator map;
+        ComponentRef iter_cref;
+        Pointer<Variable> iter_var;
+
+      // it already is a proper range, use it for the for loop
+      case (node, range as Expression.RANGE()) then (ComponentRef.makeIterator(node, Type.INTEGER()), range, NONE());
+
+      // it has an array as constructor, map it to a range
+      // used to fix #13031
+      case (node, range as Expression.ARRAY()) algorithm
+        node2   := InstNode.newIterator("$" + InstNode.name(node), Type.INTEGER(), sourceInfo());
+        range2  := Expression.makeRange(Expression.INTEGER(1), NONE(), Expression.INTEGER(Type.sizeOf(Expression.typeOf(range))));
+        map     := Iterator.fromFrames({(ComponentRef.makeIterator(node, Type.INTEGER()), range, NONE())});
+
+        // create the new iterator variable
+        iter_cref := ComponentRef.makeIterator(node2, Type.INTEGER());
+        iter_var  := BackendDAE.lowerIterator(iter_cref);
+        iter_cref := BVariable.getVarName(iter_var);
+        UnorderedSet.add(iter_var, set);
+      then (iter_cref, range2, SOME(map));
+
+      else algorithm
+        Error.addMessage(Error.INTERNAL_ERROR,{getInstanceName() + " failed to inline iterator expression: " + InstNode.toString(Util.tuple21(iter)) + " in " + Expression.toString(Util.tuple22(iter)) + "."});
+      then fail();
+    end match;
+  end inlineArrayIterator;
 
   function createInlinedEquation
     "used for inlining record, tuple and array equations.
@@ -534,6 +585,7 @@ protected
     input EquationAttributes attr;
     input Iterator iter;
     input VariablePointers variables;
+    input UnorderedSet<VariablePointer> set "new iterators";
     input Pointer<Integer> index;
   protected
     Pointer<list<Pointer<Equation>>> tmp_eqns = Pointer.create({});
@@ -541,7 +593,7 @@ protected
     Pointer<Equation> new_eqn;
   algorithm
     new_eqn := Equation.makeAssignment(lhs, rhs, index, NBEquation.SIMULATION_STR, iter, attr);
-    inlined := inlineRecordTupleArrayEquation(Pointer.access(new_eqn), iter, variables, tmp_eqns, index, false);
+    inlined := inlineRecordTupleArrayEquation(Pointer.access(new_eqn), iter, variables, tmp_eqns, set, index, false);
     eqns := match inlined
       case Equation.DUMMY_EQUATION() then listAppend(eqns, Pointer.access(tmp_eqns));
       else algorithm

--- a/OMCompiler/Compiler/NBackEnd/Modules/3_Post/NBTearing.mo
+++ b/OMCompiler/Compiler/NBackEnd/Modules/3_Post/NBTearing.mo
@@ -317,11 +317,12 @@ protected
     Tearing strict;
   protected
     list<list<Slice<EquationPointer>>> acc;
+    UnorderedSet<VariablePointer> dummy_set = UnorderedSet.new(BVariable.hash, BVariable.equalName);
   algorithm
     comp := match comp
       case StrongComponent.ALGEBRAIC_LOOP(strict = strict) algorithm
         // inline potential records
-        acc := list(Inline.inlineRecordSliceEquation(eqn, variables, eq_index, true) for eqn in strict.residual_eqns);
+        acc := list(Inline.inlineRecordSliceEquation(eqn, variables, dummy_set, eq_index, true) for eqn in strict.residual_eqns);
 
         // create residual equations
         strict.residual_eqns := list(Slice.apply(eqn, function Equation.createResidual(new = true)) for eqn in List.flatten(acc));

--- a/OMCompiler/Compiler/NBackEnd/Util/NBSlice.mo
+++ b/OMCompiler/Compiler/NBackEnd/Util/NBSlice.mo
@@ -761,7 +761,7 @@ public
           Expression linMap;
 
         // equal inertia, combine the frames
-        case ((inertia1, (loc1, (name1, _))), (inertia2, (loc2, (name2, _)))) guard(inertia1 == inertia2) algorithm
+        case ((inertia1, (loc1, (name1, _, _))), (inertia2, (loc2, (name2, _, _)))) guard(inertia1 == inertia2) algorithm
           addOp := Operator.fromClassification((NFOperator.MathClassification.ADDITION, NFOperator.SizeClassification.SCALAR), Type.INTEGER());
           mulOp := Operator.fromClassification((NFOperator.MathClassification.MULTIPLICATION, NFOperator.SizeClassification.SCALAR), Type.INTEGER());
           if arrayLength(loc1) <> arrayLength(loc2) then
@@ -1568,10 +1568,11 @@ protected
       local
         ComponentRef name;
         Expression exp;
+        Option<Iterator> map;
 
-      case (name, exp as Expression.RANGE()) then (name, Expression.sliceRange(exp, range));
+      case (name, exp as Expression.RANGE(), map) then (name, Expression.sliceRange(exp, range), map);
 
-      case (_, exp) algorithm
+      case (_, exp, _) algorithm
         Error.addMessage(Error.INTERNAL_ERROR,{getInstanceName()
           + " failed because frame expression was not Expression.RANGE(): " + Expression.toString(exp)});
       then fail();

--- a/OMCompiler/Compiler/NFFrontEnd/NFExpression.mo
+++ b/OMCompiler/Compiler/NFFrontEnd/NFExpression.mo
@@ -1178,21 +1178,6 @@ public
     output Integer start;
     output Integer step;
     output Integer stop;
-  protected
-    function getInteger
-      input Expression exp;
-      output Integer i;
-    protected
-      Expression e;
-    algorithm
-      e := Expression.map(exp, Expression.replaceResizableParameter);
-      i := match SimplifyExp.simplify(e)
-        case INTEGER(i) then i;
-        else algorithm
-          Error.assertion(false, getInstanceName() + " cannot be parsed to an integer: " + toString(exp), sourceInfo());
-        then fail();
-      end match;
-    end getInteger;
   algorithm
     (start, step, stop) := match range
       local
@@ -1217,6 +1202,21 @@ public
       then fail();
     end match;
   end getIntegerRange;
+
+  function getInteger
+    input Expression exp;
+    output Integer i;
+  protected
+    Expression e;
+  algorithm
+    e := Expression.map(exp, Expression.replaceResizableParameter);
+    i := match SimplifyExp.simplify(e)
+      case INTEGER(i) then i;
+      else algorithm
+        Error.assertion(false, getInstanceName() + " cannot be parsed to an integer: " + toString(exp), sourceInfo());
+      then fail();
+    end match;
+  end getInteger;
 
   function makeTuple
     input list<Expression> expl;

